sa/storage-authority.go

```diff
@@ -673,6 +674,117 @@ func (ssa *SQLStorageAuthority) FinalizeAuthorization(authz core.Authorization)
 	return
 }
 
+var authorizationTables = []string{
+	"authz",
+	"pendingAuthorizations",
+}
+
+const getAuthorizationIDsMax = 1000
+
+func getAuthorizationIDsByDomain(db *gorp.DbMap, ident string, now time.Time) ([]string, error) {
+	var allIDs []string
+	for _, tableName := range authorizationTables {
+		_, err := db.Select(
+			&allIDs,
+			fmt.Sprintf(
+				`SELECT id FROM %s
+         WHERE identifier = :ident AND
+         status != :invalid AND
+         status != :revoked AND
+         expires > :now
+         LIMIT :limit`,
+				tableName,
+			),
+			map[string]interface{}{
+				"ident":   ident,
+				"invalid": string(core.StatusInvalid),
+				"revoked": string(core.StatusRevoked),
+				"now":     now,
+				"limit":   getAuthorizationIDsMax,
+			},
+		)
+		if err != nil {
+			return nil, err
+		}
+		if len(allIDs) == getAuthorizationIDsMax {
+			break
+		}
+	}
+	return allIDs, nil
+}
+
+func revokeAuthorizations(db *gorp.DbMap, authIDs []string) (int64, int64, error) {
+	results := []int64{0, 0}
+	quotedIDs := []string{}
+	for _, id := range authIDs {
+		quotedIDs = append(quotedIDs, strconv.Quote(id))
+	}
+	idStr := strings.Join(quotedIDs, ", ")
+	for i, tableName := range authorizationTables {
+		result, err := db.Exec(
+			fmt.Sprintf(
+				`UPDATE %s
+         SET status = ?
+         WHERE id IN (%s)`,
+				tableName,
+				idStr,
+			),
+			string(core.StatusRevoked),
+		)
+		if err != nil {
+			return results[0], results[1], err
+		}
+		batchSize, err := result.RowsAffected()
+		if err != nil {
+			return results[0], results[1], err
+		}
+		results[i] = batchSize
+	}
+	_, err := db.Exec(
+		fmt.Sprintf(
+			`UPDATE challenges
```

Comment:
If we have the challenge revocations after the authorization revocation, failures in the middle of the authorization updates will mean that the next run of this RevokeAuthorizationsByDomain won't see the challenges that need to be revoked anymore. We'll miss those challenges because some of the authorizations will be marked as "revoked" and so won't be returned by the get function on the second run and so won't be used in here.

Larger question: why do we need revoke the challenges? I know there's got to be a reason but I've forgotten it.